Fixes #5929 - Taxonomy filter obey permissions

ares opened the pull request:
With this patch you can assign permissions like assign_organizations and
assign_locations to particular user so that they can then assign taxonomies
only from set of taxonomies granted by their filters.

ares commented:
~~Don't test yet, we may need similar change for other objects~~

dmitri-d commented on app/models/concerns/dirty_associations.rb:
```diff
@@ -0,0 +1,55 @@
+      end
+    end
+  end
+end
```
A test for this would be good.

dmitri-d replied:
nm, found tests, my apologies.

dmitri-d commented on app/models/concerns/dirty_associations.rb:
```diff
@@ -0,0 +1,55 @@
+# ActiveModel::Dirty does not support associations but it's sometime it'd be useful
+# so you could ask like this
+#   @user.role_ids_was
```
Probably worth mentioning that this doesn't detect changes of associated values, only changes to the collection itself...

ares replied:
I added a notice in comment below the example, thanks.

dmitri-d commented on app/models/concerns/taxonomix.rb:
```diff
@@ -147,6 +152,33 @@ def children_of_selected_organization_ids
+
+      if tried.empty? || !tried.subset?(allowed)
+        errors.add key, _('Invalid selection, you must select at least one of yours')
+      end
```
Worth mentioning whether it is Organization or Location the user needs to select -- it's unclear what the issue is otherwise.

dmitri-d commented on test/lib/dirty_associations_test.rb:
```diff
@@ -0,0 +1,28 @@
+    @tester.organization_ids = [3]
+    assert_equal @tester.organization_ids_was, [1, 2]
+  end
+end
```
A test that verifies DirtyAssociations handles multiple subsequent association changes would be useful, I think

ares replied:
Could you elaborate a bit more? This test is testing 2 changes (first assignment is in setup), what is the reason for adding 3rd change?

dmitri-d replied:
Hmm, I didn't catch you were tracking changes for new (not persisted) records, not sure how useful that is (I don't think `Dirty` module does that). 

I think it's worth verifying (for records that have been persisted) that when you have several changes stacked, like: 

```
@tester.organization_ids = [3]
...
@tester.organization_ids = [3, 4]
```

`@tester.organization_ids_was` will still return [1, 2]

ares replied:
That's not how it works with ActiveRecord. When you assign array of ids to a relation, it triggers save so the 'old' value changes immediately (for already persisted records). Plus the testing object knows nothing about persistence (which is good) it tests only the change tracking.

dmitri-d replied:
Oops, true, that makes sense for direct assignment. It will behave the same way though (I think, haven't tried) when organizations.build is used. Probably worth keeping the initial "was" value once it's been initialized, just to have the same semantics as the ActiveModel::Dirty. Or mention the difference in the docs? 

Sry to be difficult about this, this sort of stuff tends to trip me up...

ares replied:
Fair argument, no need to apologize, I added a comment with example. Or we could add some logic around `#persisted?` if needed or preferred.

dmitri-d commented on test/test_helper.rb:
```diff
@@ -103,6 +103,17 @@ def in_taxonomy(taxonomy)
+      result = yield
+      SETTINGS[:organizations_enabled] = org_settings
+      SETTINGS[:locations_enabled] = loc_settings
+      result
```
restoring SETTINGS values should probably be moved to `ensure` block, as yield may throw an exception (failed assert, for example).

ares replied:
Good point, fixed. With latest version we don't need this helper but I think it can be useful later so if you don't mind I'd keep it.

ares committed:
Fixes #5929 - Taxonomy filter obey permissions

With this patch you can assign permissions like assign_organizations and
assign_locations to particular user so that they can then assign
taxonomies
only from set of taxonomies granted by their filters.

Global users would be still able to assign any taxonomy to a resource as
long as they have appropriate assign permissions. They can also leave
the resource global.

dmitri-d commented:
lgtm.

domcleal commented on app/models/concerns/taxonomix.rb:
```diff
@@ -147,6 +152,34 @@ def children_of_selected_organization_ids
+      assoc = assoc_base.pluralize
+      key = assoc_base + '_ids'
+
+      next if (User.current.nil? || User.current.send("#{assoc}").empty?) || (!new_record? && !self.send("#{key}_changed?"))
```
Why skip if the user isn't associated to any?

ares replied:
If user is not restricted to any taxonomy (let's say org in this case) it means he can assign any organization. Imagine the opposite, global user (without taxonomy restriction) would be allowed to assign limited set of organizations. In that case all such users should probably have to be assigned all organizations otherwise they wouldn't be able to work with objects which they can see and have permissions to modify. This would be huge change for existing installations. Also it seems consistent with the fact that object assigned to no taxonomy is considered global.

domcleal commented:
Is the note about "Don't test yet, we may need similar change for other objects" still valid?

@isratrade please review and test this, thanks.

ares commented:
Feel free to test, I updated the comment

ares commented:
ping @isratrade or whoever can find some time to test

isratrade commented:
@ares, I'm OK with merging this, but why did you call the method `ensure_taxonomies_not_escalated`? What do you mean by 'escalated'?

ares commented:
It was meant to be consistent with admin flag and roles escalation. Escalation comes from privileges escalation, user could get permissions to work with objects in more taxonomies that was desired.

ares commented:
@isratrade is there anything blocking this PR?

ares commented:
or anyone else? maybe @witlessbird since you already saw the code

ares commented:
ping @isratrade

isratrade commented:
@ares, I tested this manually in the console for the following and all works as expected.
1) adding and removing locations and organizations from a domain
domain.location_ids_changed?
domain.location_ids_was
domain.organization_ids_changed?
domain.organization_ids_was
2) adding and removing locations and organizations from a user
user.location_ids_changed?
user.location_ids_was
user.organization_ids_changed?
user.organization_ids_was
3) adding and removing roles from a user
user.role_ids_changed?
user.role_ids_was

isratrade commented:
@ares, I tested in the UI and see that the multiple select for locations is blank if a non-admin user does **not** have privilege `assigned_locations`. It's debatable from a UI point of view, whether to hide the locations tab, rather than having the multiple select options blank. However, when I save a domain with the locations tab blank, I get a failure on `ensure_taxonomies_not_escalated`

The log says

```
Failed to save: Location ids Invalid locations selection, you must select at least one of yours
```

ares commented:
Thanks @isratrade, changing labels accordingly.

isratrade commented:
@ares, I thought that the PUT for `location_ids` might be blank, but the log shows that it `location_ids` is populated

ares commented:
so you need to set two permissions, view_locations let you see the view_location tab, assign_location gives you ability to change the assignment, but it seems you found an issue in corner case when no taxonomy is selected, I'll have to dig a bit deeper, it should probably unassign all locations in case you've described, right?

ares commented:
ah I was wrong, this is expected, if you try to unassign all locations, you'd make the resource available in all locations, but since you're assigned to specific locations, you can't do that, you'd make it available somewhere where you don't have permissions
does it make sense @isratrade?

ares commented:
rebased, setting label back to need testing

isratrade commented:
@ares, it makes sense, but when testing locally, the current PR doesn't allow a non-admin user with **no** permission to `assign_locations` to update a domain. There shouldn't be a validation error, but it there is, it should appear in the UI and not only in the logs which is what I experienced.

ares commented:
So it turned out that we don't display the validation, because multiple select can't display it. We could workaround by adding errors to base which is not ideal. I created a redmine issue for this http://projects.theforeman.org/issues/7319 since it's out of scope of this PR.

isratrade commented:
@ares, so it appears the validation issue I ran into is an edge case where a user has permission to `view_locations` and thus sees the locations tab, but doesn't have permission to `assign_locations` so nothing is populated in the multiple select, right? I'll add a comment to issue #7319.

I'm OK with merging this now.

ares commented:
@isratrade that's correct, unfortunately we'll encounter same behaviour when user has both permissions and tries to disassociate all locations which is not that "corner"

isratrade commented:
@ares, I verified that it works if the non-admin user does not have view_locations and assign_locations permissions. I will merge this and hopefully issue #7319 can be fixed soon.

isratrade commented:
Merged as 952396005dadab63a81cb4763e0a3dab3c3f204f. Thanks @ares !
